New emerging score - Risk Matrix

rob-taggart opened the pull request:
Please work through the following checklists. Delete anything that isn't relevant.
## Development for new xarray-based metrics
- [x] Works with n-dimensional data and includes `reduce_dims`, `preserve_dims`, and `weights` args.
- [x] Typehints added
- [x] Docstrings complete and followed Napoleon (google) style
- [x] Reference to paper/webpage is in docstring
- [x] Add error handling
- [x] Imported into the API

## Testing of new xarray-based metrics
- [x] 100% unit test coverage
- [x] Test that metric is compatible with dask.
- [x] Test that metrics work with inputs that contain NaNs
- [x] Test that broadcasting with xarray works
- [x] Test both reduce and preserve dims arguments work
- [x] Test that errors are raised as expected
- [x] Test that it works with both `xr.Dataarrays` and `xr.Datasets`

## Tutorial notebook 
- [x] Short introduction to why you would use that metric and what it tells you
- [x] A link to a reference
- [x] A "things to try next" section at the end
- [x] Add notebook to [Tutorial_Gallery.ipynb](https://github.com/nci/scores/blob/develop/tutorials/Tutorial_Gallery.ipynb)
- [ ] Optional - a detailed discussion of how the metric works at the end of the notebook

## Documentation
- [x] Add the score to the [API documentation](https://github.com/nci/scores/blob/develop/docs/api.md)
- [x] Add the score to the [included list of metrics and tools](https://github.com/nci/scores/blob/develop/docs/included.md)

rob-taggart commented:
To do:

- examples in docstring
- formulae in docstring
- datasets in tests, plus update type hints
- dask

rob-taggart commented:
@nicholasloveday , this PR is ready for review now, including a tutorial. I've uploaded a version of the paper, which explains the score. Happy to talk you through it before you review so you know the salient bits.

[warning_verification.pdf](https://github.com/user-attachments/files/17627929/warning_verification.pdf)

rob-taggart commented:
@nicholasloveday , one thing I didn't do is link the tutorial to the API. Will need to address this in the review

nicholasloveday reviewed (changes requested):
This looks good Rob. I think that the tutorial will be key in helping people adopt this score

I only have minor suggestions

nicholasloveday commented:
> @nicholasloveday , one thing I didn't do is link the tutorial to the API. Will need to address this in the review

You can add the link in https://github.com/nci/scores/blob/develop/tutorials/Tutorial_Gallery.ipynb too

and update 
https://github.com/nci/scores/blob/develop/docs/included.md

Just follow the patterns of what's already in those files

rob-taggart commented:
I've responded to every comment of your review @nicholasloveday 

@tennlee , ready for you to review/approve

It will be updated with an online URL to the paper once we submit

tennlee commented:
I'll review this shortly. Please rebase this PR has there has been quite a lot of change on the develop branch.

tennlee commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,478 @@
+"""
```
Agree, please rename this to risk_matrix.py

tennlee commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,478 @@
+from scores.utils import gather_dimensions
+
+
+def risk_matrix_score(
```
I would suggest putting a "*" after prob_threshold_dim. The purpose of this is to avoid people mucking up the order of their arguments. By forcing optional arguments to be keyword-only, certain programmer errors can be avoided.

rob-taggart replied:
done as Tennessee suggested

tennlee commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,478 @@
+    severity_dim: str,
+    prob_threshold_dim: str,
+    threshold_assignment: Optional[str] = "lower",
+) -> xr.DataArray:
```
Add a "*" before threshold_assignment

rob-taggart replied:
@nicholasloveday , I've made your suggested change.

@tennlee , I decided not to add a "*" here, because for `risk_matrix_score` you recommended to add a * after threshold_assignment. It seems odd to add one before threshold_assignment in  `_risk_matrix_score`

But will follow your direction on this if you disagree, as adding *s is new to me.

tennlee replied:
Adding the "*" has an interesting effect. It requires that the keyword argument that follow must strictly be supplied as keyword arguments. In many ways, I wonder if the best practise would be to make every argument a keyword argument, and all following an "*", which would basically amount to supplying all arguments as a dictionary.

The alternative is that keyword arguments may be supplied as keywords, or, alternatively, using positional ordering. Consider a function `def fn(a, b, c=None, d=None): ...`

This could usually be called either with

`fn(1,2)
fn(1,2,4,3)
fn(1,2,4,d=3)
fn(1,2,(c=4, d=3)`

It's very easy to mix up the ordering of positionally inferred arguments. This is less error-prone for commonly patterned arguments (e.g. fcst, obs), and more so for arguments where the ordering is essentially arbitrary. Using the "*" would result in errors for the user when using positional ordering to supply keyword-arguments.

So, in short, the pattern in `scores` is to put a "*" before the first keyword argument to help cut down on programmer mistakes which would otherwise too-easily slip through.

It is also particularly important for the user-facing API and less so for internal code, but since it's generally better for coding accuracy all around, I think it should just be done everywhere.

I hope that helps explain.

rob-taggart replied:
Thanks @tennlee , I've changed both functions so that there is a * before the kwargs

tennlee commented:
This is a nicely done pull request, well done! 

I have pushed this work to readthedocs. See:
 - https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Tutorial_Gallery.html#Emerging
 - https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Risk_Matrix_Score.html
 - https://scores.readthedocs.io/en/228-documentation-testing-branch/api.html#scores-emerging
 - https://scores.readthedocs.io/en/228-documentation-testing-branch/included.html#emerging

Please take a look at the built documentation and confirm you are happy with it.

I made the following changes:
 - Rebased the branch (you will need to update your local checkout)
 - Added admonitions to the docstrings (shows in the readthedocs as a box that says "caution")
 - Tagged the 'emerging' tutorial so it shows consistently in the tutorial gallery
 - Updated the layout of "included.md" for consistency with other similar groups of functions
 - Silenced some mypy errors which didn't have an easy solution as they are intersection types of iterable and sized

Steph-Chong commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,490 @@
+        This function is part of an emerging score that is under development and has not yet
+        been peer-reviewed. It may change at any time.
+
+    Generates a 2-dimensional xr.DataArray of the decision thresholds for the risk matrix
+    score calculation.  Assumes that values toward the left in ``matrix_weights`` correspond
```
```suggestion
    Generates a 2-dimensional xr.DataArray of the decision thresholds for the :py:func:`scores.emerging.risk_matrix_score` calculation.  Assumes that values toward the left in ``matrix_weights`` correspond
```

Steph-Chong replied:
@tennlee accepted this change into the PR

Steph-Chong commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,490 @@
+        been peer-reviewed. It may change at any time.
+
+    Given a warning scaling matrix, assessment weights and other inputs,
+    returns the decision weights for the risk matrix score as an xarray data array.
```
```suggestion
    returns the decision weights for the :py:func:`scores.emerging.risk_matrix_score` as an xarray data array.
```

Steph-Chong replied:
@tennlee accepted this change into the PR

Steph-Chong reviewed (commented):
Hi @rob-taggart - I help out with documentation.

Just a heads up, I often do my reviews in batches, so may provide some more feedback later.

tennlee commented on src/scores/categorical/multicategorical_impl.py:
```diff
@@ -3,7 +3,7 @@
 
 from collections.abc import Sequence
-from typing import Optional, Union
+from typing import Iterable, Optional, Union
```
I have removed this, it is not needed.

Steph-Chong commented:
@rob-taggart , in @nicholasloveday's review above, GitHub has "hidden" 17 of Nick's comments.

In order to see his review comments, you need to click on the box that says "17 hidden converations Load more..." (see screenshot below of this box).

![image](https://github.com/user-attachments/assets/8d772beb-c8e9-416c-bb7b-c60f12e6c7ab)

Steph-Chong commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2690 @@
+   "metadata": {},
+   "source": [
+    "# The Risk Matrix Score\n",
+    "\n",
+    "The Risk Matrix Score provides a consistent way of scoring forecasts and warnings within a risk matrix framework (Taggart & Wilke 2024). A lower score is better.\n",
```
```suggestion
    "\n",
    "\n",
    "> **CAUTION:**\n",
    "> This is an emerging score that is under development and has not yet been peer-reviewed. It may change at any time.
    "\n",
    "The Risk Matrix Score provides a consistent way of scoring forecasts and warnings within a risk matrix framework (Taggart & Wilke 2024). A lower score is better.\n",
```
This change adds an admonition to the tutorial (that renders acceptably in both JupyterLab and in Read the Docs)

Steph-Chong replied:
Below is a screenshot of how it appeared when I built it locally in Read the Docs (it looked very similar in JupyterLab)

![image](https://github.com/user-attachments/assets/736bf824-2d23-40f8-83ff-f599df423767)

Steph-Chong replied:
@tennlee accepted this change into this PR.

Steph-Chong commented on docs/included.md:
```diff
@@ -739,4 +739,32 @@
+  - [Tutorial](project:./tutorials/Risk_Matrix_Score.md)
+  - Taggart, R. J., & Wilke, D. J. (2024). Consistent evaluation of warnings that are based on risk matrices. In preparation.
+* -  
+    - Risk Matrix Score (Matrix Weights To Array)
```
@tennlee accepted this change into the PR.

Steph-Chong commented on docs/included.md:
```diff
@@ -739,4 +739,32 @@
+  - [Tutorial](project:./tutorials/Risk_Matrix_Score.md)
+  - Taggart, R. J., & Wilke, D. J. (2024). Consistent evaluation of warnings that are based on risk matrices. In preparation.
+* -  
+    - Risk Matrix Score (Scaling to Weight Array)
```
@tennlee accepted this change into the PR

rob-taggart commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2691 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that the weight for the (MOD+, 0.5) decision point is 0. This is because the warning level doesn't change at this decision point. That is, if one is only interested in the final warning level outcome, discriminating above 0.5 probability threshold and below 0.5 probability threshold for the MOD+ severity category is inconsequential.\n",
```
done

rob-taggart commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,478 @@
+    threshold_assignment: Optional[str] = "lower",
+) -> xr.DataArray:
+    """
+    Calculates the risk matrix score MS of Taggart and Wilke (2024).
```
"MS" is the abbreviation given to "Matrix Score" in the paper. In contrast to the warning score (WS) in the paper.

I'll leave it as is now, but it might need updating prior to the paper being submitted. We are thinking of using the abbreviation RMS instead.

Steph-Chong replied:
Given "MS" is an abbreviation, and it is the first time the abbreviation is being mentioned, I agree that it would be clearer to include it here in brackets - i.e. "(MS)".

Similarly, if down the road it is changed to RMS, I would suggest including it in brackets as "(RMS)" when it is first introduced as the abbreviation.

rob-taggart replied:
OK, I'll go along with the suggestion. I was using mathematics paper writing conventions, and in this case one doesn't require parentheses if adhering to those conventions. This context is different I suppose.

tennlee commented on src/scores/emerging/__init__.py:
```diff
@@ -1,5 +1,5 @@
 """
-This module is currently empty.
+Import the functions from the implementations into the public API.
```
I think this line can just be deleted

rob-taggart replied:
Looks like this suggestion has already been committed into the branch

Steph-Chong reviewed (commented):
Hi @rob-taggart - I read through your tutorial in Read the Docs (on branch 228) and the GitHub diff (but did not run it this time in JupyterLab). 

I have left some comments with very minor feedback.

rob-taggart commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2691 @@
+{
```
I've added the following to one of the paragraphs in the intro

> We now set up a synthetic experiment which will generate forecasts from three different forecasters and corresponding observations. We will score the forecasters using the Risk Matrix Score in two different ways: (1) use a choice of decision weights that evaluates how they fill out the risk matrix, and (2) use a choice of decision weights that evaluates the warning levels they issue.

rob-taggart commented on src/scores/emerging/emerging_impl.py:
```diff
@@ -0,0 +1,478 @@
+    return weight_matrix
+
+
+def scaling_to_weight_array(
```
Changed to `weights_from_warning_scaling`

rob-taggart commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2694 @@
+    "\n",
+    "## Things to try next\n",
+    "\n",
+    "Try adding a two new forecasters to the mix. The first new forecaster doubles SynopticSally's probability forecasts (clipped at 1), while the second new forecaster halves SynopticSally's probability forecasts. How do you think these new forecasters will rank (using the risk matrix score) compared with SeasonalSam? Can you find two different `decision_weights` that result in different ranking ...
```
thanks , fixed

rob-taggart commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2694 @@
+    "\n",
+    "This experiment is repeated independently 10000 times.\n",
+    "\n",
+    "We will now implement this in python. In order to calculate the risk matrix score:\n",
```
done

rob-taggart commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2694 @@
+   "source": [
+    "A lower score is better, so SynopticSally has done best overall, followed by SeasonalSam then NeverWarnNate. This is to be expected because SynopticSally makes predictions based on better information than the other forecasters.\n",
+    "\n",
+    "We now calculate at the risk matrix score, but where we only put weights on decision points that affect the final warning level outcome. To do this, we need to generate the weight matrix based on the warning scaling (i.e., the colouring scheme in the risk matrix) and on supplied assessment weights. Assessment weights indicate the importance of discriminating between the different warning lev ...
```
done

rob-taggart commented on tutorials/Risk_Matrix_Score.ipynb:
```diff
@@ -0,0 +1,2694 @@
+    "\n",
+    "## Reference\n",
+    "\n",
+    "Taggart, R. J. & Wilke, D. J., Consistent evaluation of warnings that are based on risk matrices (2024). In preparation."
```
Thanks. Fixed, plus updated the title of the paper here and in risk_matrix.py

rob-taggart commented:
@nicholasloveday , @tennlee , @Steph-Chong , thanks so much for your reviews. I believe I have addressed all your feedback. Please let me know if there is anything else I need to do

Steph-Chong commented on docs/included.md:
```diff
@@ -739,4 +739,32 @@
+    - Risk Matrix Score - Matrix Weights To Array
+  - [API](api.md#scores.emerging.matrix_weights_to_array)
+  - [Tutorial](project:./tutorials/Risk_Matrix_Score.md)
+  - Taggart, R. J., & Wilke, D. J. (2024). Consistent evaluation of warnings that are based on risk matrices. In preparation.
```
I think this needs to be updated to the new paper title.

rob-taggart replied:
Ta

Steph-Chong commented on docs/included.md:
```diff
@@ -739,4 +739,32 @@
+    - Risk Matrix Score - Warning Scaling to Weight Array
+  - [API](api.md#scores.emerging.weights_from_warning_scaling)
+  - [Tutorial](project:./tutorials/Risk_Matrix_Score.md)
+  - Taggart, R. J., & Wilke, D. J. (2024). Consistent evaluation of warnings that are based on risk matrices. In preparation.
```
I think this needs to be updated to the new paper title.

Steph-Chong commented on src/scores/emerging/risk_matrix.py:
```diff
@@ -0,0 +1,497 @@
+    returns the decision weights for the :py:func:`scores.emerging.risk_matrix_score` as an xarray data array.
+
+    Comprehensive checks are made on ``scaling_matrix`` to ensure it satisfies the properties
+    of warning scaling in Table 1 of Taggart & Wilke (2004).
```
```suggestion
    of warning scaling in Table 1 of Taggart & Wilke (2024).
```
Is this meant to be Taggart & Wilke (2024)?

Steph-Chong reviewed (commented):
@rob-taggart I have made a few comments about the references. (Some minor typos and the references in included.md need their titles updated).

Steph-Chong commented:
I also have one question. In risk_matrix.py, there are five occurrences of the phrase "xarray data array" (lines 91, 238, 291, 359, 382). Should any (or all) of them be "xarray.DataArray"? 

I honestly have no opinion on this matter, I just wanted to raise the question for your consideration. If "xarray data array" is the most appropriate choice, then no worries.

rob-taggart commented:
> I also have one question. In risk_matrix.py, there are five occurrences of the phrase "xarray data array" (lines 91, 238, 291, 359, 382). Should any (or all) of them be "xarray.DataArray"?
> 
> I honestly have no opinion on this matter, I just wanted to raise the question for your consideration. If "xarray data array" is the most appropriate choice, then no worries.

Thanks @Steph-Chong . I don't have an opinion either. I was following the example of docstrings from some other metrics in scores. Nonetheless, I updated "xarray data array" to "xarray object" in two cases as the code had since been updated to handle (and hence return) xr.Dataset as well as xr.DataArray. In the other instances only xr.DataArray is returned and I left as is.

rob-taggart commented:
Thanks for your recent review @Steph-Chong , all recommended changes have been made (so I believe)

Steph-Chong commented on docs/api.md:
````diff
@@ -109,3 +109,9 @@
+```{eval-rst}
+.. autofunction:: scores.emerging.risk_matrix_score
+.. autofunction:: scores.emerging.matrix_weights_to_array
+.. autofunction:: scores.emerging.scaling_to_weight_array
````
```suggestion
.. autofunction:: scores.emerging.weights_from_warning_scaling
```
Is this mean to be weights_from_warning_scaling?

Steph-Chong replied:
(Sorry, I didn't notice this when I was doing my previous round of review)

tennlee commented:
Hi @rob-taggart, I have updated branch 228 on readthedocs with the latest work.

See:

- [ ] https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Tutorial_Gallery.html#Emerging
- [ ] https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Risk_Matrix_Score.html
- [ ] https://scores.readthedocs.io/en/228-documentation-testing-branch/api.html#scores-emerging
- [ ] https://scores.readthedocs.io/en/228-documentation-testing-branch/included.html#emerging

Please take a look at the built documentation and check off each list item above to confirm you are happy with it.

tennlee commented:
This looks good to me. Once you have ticked off the final documentation review, I will go ahead and merge this.

Steph-Chong commented:
@rob-taggart Below is a screenshot from the docstrings rendered in Read the Docs on branch 228.

![image](https://github.com/user-attachments/assets/0ef4ba4e-617c-4fb2-acc6-cfe216710124)

If this is how you want your references to render, no worries. 

But if the bold text was unintentional - it was caused by putting "In preparation" on a new line with a tab/spaces before the "In preparation". If you instead move "In preparation" back to the same line as the paper title (with just one space between "evaluation." and "In"), then it will render as below:

![image](https://github.com/user-attachments/assets/268603c7-0f7f-4285-9248-e14b572a7778)

(Note: when I ran black locally, black didn't get angry about the line length).

rob-taggart commented:
> @rob-taggart Below is a screenshot from the docstrings rendered in Read the Docs on branch 228.
> 
> ![image](https://private-user-images.githubusercontent.com/168800785/391388551-0ef4ba4e-617c-4fb2-acc6-cfe216710124.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMxMDcwOTksIm5iZiI6MTczMzEwNjc5OSwicGF0aCI6Ii8xNjg4MDA3ODUvMzkxMzg4NTUxLTBlZjRiYTRlLTYxN2MtNGZiMi1hY2M2LWNmZTIxNjcxMDEyNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMlQwMjMzMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01N2JlNDg0NjFmOTdmYzQ3NDJjZDYyMDc0YzJiZjE0YTRhM2MyYWMxNzk3ZmM0ZTlkNWYyZDY4NDE4Nzc3MDdkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pMbwQE85Cb0x9ImrI6lF-L81aGKRC33dU367Sh1VcwY)
> 
> If this is how you want your references to render, no worries.
> 
> But if the bold text was unintentional - it was caused by putting "In preparation" on a new line with a tab/spaces before "In preparation". If you instead move "In preparation" back to the same line as the paper title (with just one space between "evaluation." and "In"), then it will render as below:
> 
> ![image](https://private-user-images.githubusercontent.com/168800785/391390106-268603c7-0f7f-4285-9248-e14b572a7778.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMxMDcwOTksIm5iZiI6MTczMzEwNjc5OSwicGF0aCI6Ii8xNjg4MDA3ODUvMzkxMzkwMTA2LTI2ODYwM2M3LTBmN2YtNDI4NS05MjQ4LWUxNGI1NzJhNzc3OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMlQwMjMzMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NDk1NzMyZGY4YTU2OTBlYThkODhjZWNiNjIxYmZhYWQ5ZGYwMDZkM2E1ZjM2YmVmNjg3Zjk2NmU0OTRmODY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2NngMFoAjdcxR_e0tld-43ohxDlN8LCD4qyUur0d3c0)
> 
> (Note: when I ran black locally, black didn't get angry about the line length).

Thanks @Steph-Chong , yeah, bold text wasn't my intension so I've fixed it up as you suggested

rob-taggart commented:
> Hi @rob-taggart, I have updated branch 228 on readthedocs with the latest work.
> 
> See:
> 
> * [ ]  https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Tutorial_Gallery.html#Emerging
> * [ ]  https://scores.readthedocs.io/en/228-documentation-testing-branch/tutorials/Risk_Matrix_Score.html
> * [ ]  https://scores.readthedocs.io/en/228-documentation-testing-branch/api.html#scores-emerging
> * [ ]  https://scores.readthedocs.io/en/228-documentation-testing-branch/included.html#emerging
> 
> Please take a look at the built documentation and check off each list item above to confirm you are happy with it.

Thanks @tennlee , I had a read of everything. I've updated the docstrings so that "matrix decision thresholds" is now replaced with "decision points in the risk matrix" (or similar). This is more consistent with the latest version of the paper (which is almost ready for submission), and is better terminology, since each decision point is (typically) a combination of thresholds, not just a single threshold.

So if you and @Steph-Chong are happy with these changes, I'm happy to have it merged

Steph-Chong commented on src/scores/emerging/risk_matrix.py:
```diff
@@ -298,11 +298,10 @@ def matrix_weights_to_array(
 
     Examples:
-        Returns risk matrix decision weights, where weights increase with increasing
+        Returns weights for eachh risk matrix decision point, where weights increase with increasing
```
```suggestion
        Returns weights for each risk matrix decision point, where weights increase with increasing
```

rob-taggart commented:
Thanks for the typo pick up @Steph-Chong , I've committed your suggested change

rob-taggart commented:
thanks for your reviews, @tennlee , @Steph-Chong , @nicholasloveday , much appreciated
